icons.go

```diff
@@ -0,0 +1,212 @@
+		}
+
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				log.TErrorf("Failed to close response body, error: %s", err)
+			}
+		}()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read respnse body, error: %s", err)
+		}
+
+		if resp.StatusCode != http.StatusCreated {
+			return fmt.Errorf("invalid status code: %d, headers: %s, body: %s", resp.StatusCode, resp.Header, body)
+		}
+
+		decoded := map[string][]appIconCandidateURL{
+			"data": uploadURLs,
+		}
+
+		if err = json.Unmarshal(body, &decoded); err != nil {
+			return fmt.Errorf("failed to unmarshal resoponse body, error: %s", err)
+		}
+		uploadURLs = decoded["data"]
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to upload, error: %s", err)
+	}
+	return uploadURLs, nil
+}
+
+func uploadIcon(filePath string, iconCandidate appIconCandidateURL) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to open file (%s), error: %s", filePath, err)
+	}
+	defer func() {
+		if err := file.Close(); err != nil {
+			log.TWarnf("failed to close file, error: %s", err)
+		}
+	}()
+
+	// If a byte array is passed to http.NewRequest, the Content-lenght header is set to its lenght.
+	// That does not seem apply to a stream (as it has no defined lenght).
+	// The Content-lenght header is signed by S3, so has to match to the filesize sent
+	// in the getUploadURL() function.
+	data, err := ioutil.ReadAll(file)
+	if err != nil {
+		return fmt.Errorf("can not read file, error: %s", err)
+	}
+	if int64(len(data)) != iconCandidate.FileSize {
+		return fmt.Errorf("array lenght deos not match to file size reported to the API, "+
+			"actual: %d, expected: %d",
+			len(data), iconCandidate.FileSize)
+	}
+
+	if err := retry.Times(3).Wait(5 * time.Second).Try(func(attemp uint) error {
+		if attemp != 0 {
+			log.TWarnf("%d query attemp failed", attemp)
+		}
+
+		request, err := http.NewRequest(http.MethodPut, iconCandidate.UploadURL, bytes.NewReader(data))
+		if err != nil {
+			return fmt.Errorf("failed to create request, error: %s", err)
+		}
+
+		request.Header.Add("Content-Type", "image/png")
+
+		resp, err := http.DefaultClient.Do(request)
+		if err != nil {
+			return fmt.Errorf("failed to submit, error: %s", err)
+		}
+
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				log.TErrorf("Failed to close response body, error: %s", err)
+			}
+		}()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read respnse body, error: %s", err)
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			return fmt.Errorf("invalid status code: %d, headers: %s, body: %s", resp.StatusCode, resp.Header, body)
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("failed to upload, error: %s", err)
+	}
+	return nil
+}
+
+func validateIcon(iconPath string) error {
+	const maxImageSize = 1024
+	const maxFileSize = 2 * 1e6
```

Comment:
This makes me inconvenient. Is there any other way to setup a binary file size? For example 1024*smtg?